co-log/src/Colog/Concurrent.hs

```diff
@@ -0,0 +1,129 @@
+-- |
+-- Sometimes an application may need to dump logs asynchonously.
+-- This may happen if the target where logs are dumped is
+-- slow and logs should be dumped in chunks, or if application
+-- should split CPU and I/O-bound tasks. In these cases it
+-- may make sense to run asynchronous logger.
+--
+-- Asynchronous logger runs in background thread, so users
+-- thread may dump logs asynchronously.
+--
+-- This module provides basic interface for the concurrent
+-- logging that may be either used as-is or can be used to
+-- build more complex concurrency abstractions.
+module Colog.Concurrent
+       ( -- $worker-thread
+         mkBackgroundThread
+       , runInBackgroundThread
+       -- $background-worker
+       , BackgroundWorker
+       , forkBackgroundLogger
+       , killBackgroundLogger
+       , convertToLogAction
+       ) where
+
+import Control.Concurrent
+import Control.Concurrent.STM.TBQueue
+
+import Colog.Core.Action (LogAction (..))
+import Colog.Concurrent.Internal
+
+-- $worker-thread
+-- Worker thread approach allows to create a single worker
+-- thread that will perform all logging. This approach allows
+-- user to use different loggers and change them on the flight.
+--
+-- With bracket resource control style code may look as
+-- following:
+--
+-- @
+-- bracket ('mkBackgroundThread' 4096)
+--         ('killBackgroundThread')
+--         $ \thread -> 
+--            let la1 = runInBackgroundThread worker logger1
+--                la2 = runInBackgroundThread worker logger1
+--            unLogger la1 msg1
+--            unLogger la2 msg2
+--            ..
+-- @
+--
+-- This code will run a background thread and process different
+-- loggers in that thread.
+--
+-- You may extend the logger and entire logger will be run
+-- on the background thread. It is not a case with
+-- 'forkBackgroundLogger'.
+
+-- | Create a background worker with a given capacity.
+-- If capacity is reached, then the thread that tries to
+-- write logs will be blocked.
+mkBackgroundThread :: Int -> IO (BackgroundWorker (IO ()))
```

Comment:
Input message as `IO ()` looks interesting 🤔 I wonder how useful such simple `LogAction` could be:

```haskell
idLogAction :: LogAction m (m ())
idLogAction = LogAction id
```

Reply:
I'm not very good at CT to find what that object will be and derive a use-case for that based on that knowledge. Right now I don't see how it can be used :/